### PR TITLE
Use rabbitmq_delayed_message_exchange plugin

### PR DESCRIPTION
**Pull request by cpetschnig**
Work in progress

### CODE REVIEW

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb [hunk 1]
```diff
@@ -32,51 +34,83 @@ def register_event(event_type)
...
+          exchange = @queue_manager.get_exchange(channel, @event_raised_exchange)
 
-          channel.close
-          connection.close
-          raise e
-        end
+          delay_exchange = @queue_manager.get_delay_exchange(channel, queue)
+
+          delay_queue = @queue_manager.create_delay_queue(channel, queue, exchange.name, delay)
+          delay_queue.bind(delay_exchange)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb]: I would add the routing key to this bind method call:

`delay_queue.bind(delay_exchange, :routing_key => _event_type)`

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb [hunk 2]
```diff
@@ -42,6 +42,11 @@ def get_subscriber_exchange(channel, queue)
         return channel.fanout("#{_queue_name}.ex")
       end
 
+      def get_delay_exchange(channel, queue)
+        _queue_name = EventQ.create_queue_name(queue.name)
+        channel.fanout("#{_queue_name}.d.ex")
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb]: I think you will need to add the delay length to the exchange name:

`channel.fanout("#{_queue_name}.#{delay}.d.ex")`

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb [hunk 3]
```diff
@@ -42,6 +42,11 @@ def get_subscriber_exchange(channel, queue)
         return channel.fanout("#{_queue_name}.ex")
       end
 
+      def get_delay_exchange(channel, queue)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb]: Add delay as a method argument:

`def get_delay_exchange(channel, queue, delay=0)`

**Review by vaughanbrittonsage — COMMENTED** [review]
See comments

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb [hunk 4]
```diff
@@ -32,51 +34,83 @@ def register_event(event_type)
...
-        ex.publish(message, :routing_key => _event_type)
-        rescue => e
+        with_connection do |channel|
+          exchange = @queue_manager.get_exchange(channel, @event_raised_exchange)
 
-          channel.close
-          connection.close
-          raise e
-        end
+          delay_exchange = @queue_manager.get_delay_exchange(channel, queue)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb]: pass `delay` into the get_delay_exchange method call:

`delay_exchange = @queue_manager.get_delay_exchange(channel, queue, delay)`

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb [hunk 5]
```diff
@@ -32,51 +34,83 @@ def register_event(event_type)
...
+      def raise_event_in_queue(event_type, event, queue, delay)
+        register_event(event_type)
 
-        message = serialization_provider.serialize(qm)
+        _event_type = EventQ.create_event_type(event_type)
 
-        ex.publish(message, :routing_key => _event_type)
-        rescue => e
+        with_connection do |channel|
+          exchange = @queue_manager.get_exchange(channel, @event_raised_exchange)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_eventq_client.rb]: replace `@queue_manager.get_exchange`with a call to `@queue_manager.get_queue_exchange`

File: eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb [hunk 6]
```diff
@@ -42,6 +42,11 @@ def get_subscriber_exchange(channel, queue)
         return channel.fanout("#{_queue_name}.ex")
       end
 
+      def get_delay_exchange(channel, queue, delay)
+        _queue_name = EventQ.create_queue_name(queue.name)
+        channel.fanout("#{_queue_name}.#{delay}.d.ex")
+      end
+
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/lib/eventq_rabbitmq/rabbitmq_queue_manager.rb]: create method `get_queue_exchange` which should create an exchange for single usage by a queue.

exchange name: `"#{_queue_name}.ex"`

File: eventq_rabbitmq/spec/eventq_rabbitmq/rabbitmq_eventq_client_spec.rb [hunk 7]
```diff
@@ -6,92 +6,139 @@
...
+    end
 
-      qm = nil
+    context 'two events with different delays' do
+      let(:other_delay_seconds) { 7 }
+      let(:other_message) { 'Brave New World' }
 
-      puts '[QUEUE] waiting for message...'
+      it 'should raise an event object with a delay' do
+        subscription_manager.subscribe(event_type, subscriber_queue)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/spec/eventq_rabbitmq/rabbitmq_eventq_client_spec.rb]: subscription should no longer be required after above changes to exchange.

File: eventq_rabbitmq/spec/eventq_rabbitmq/rabbitmq_eventq_client_spec.rb [hunk 8]
```diff
@@ -6,92 +6,139 @@
...
+    let(:queue_in) do
+      EventQ::Queue.new.tap do |queue|
+        queue.name = queue_name
+      end
     end
+    let(:delay_seconds) { 3 }
 
-    it 'should raise an event object and be broadcast to a subscriber queue' do
+    it 'should raise an event object with a delay' do
+      subscription_manager.subscribe(event_type, subscriber_queue)
```
Comment (vaughanbrittonsage) [inline eventq_rabbitmq/spec/eventq_rabbitmq/rabbitmq_eventq_client_spec.rb]: subscription should no longer be required after above changes to exchange.

**Comment by cpetschnig** [comment]
I've tested this in my application and it works as expected!

**Review by vaughanbrittonsage — APPROVED** [review]

**Pull request merged**